source/dem/non_uniform_insertion.cc

```diff
@@ -62,25 +66,43 @@ NonUniformInsertion<dim>::insert(
         Utilities::MPI::all_gather(communicator, my_bounding_box);
 
       // Distbuting particles between processors
-      if (Utilities::MPI::this_mpi_process(communicator) !=
-          (Utilities::MPI::n_mpi_processes(communicator) - 1))
+      if (this_mpi_process != (n_mpi_process - 1))
         this->inserted_this_step_this_proc =
-          floor(this->inserted_this_step /
-                Utilities::MPI::n_mpi_processes(communicator));
+          floor(this->inserted_this_step / n_mpi_process);
       else
         this->inserted_this_step_this_proc =
           this->inserted_this_step -
-          (Utilities::MPI::n_mpi_processes(communicator) - 1) *
-            floor(this->inserted_this_step /
-                  Utilities::MPI::n_mpi_processes(communicator));
+          (n_mpi_process - 1) * floor(this->inserted_this_step / n_mpi_process);
 
       // Finding insertion points using assign_insertion_points function
-      std::vector<Point<dim>> insertion_points;
-      insertion_points.reserve(this->inserted_this_step_this_proc);
-
-      this->assign_insertion_points(insertion_points,
-                                    dem_parameters.insertion_info,
-                                    communicator);
+      std::vector<Point<dim>> global_insertion_points;
+      std::vector<Point<dim>> insertion_points_on_proc;
+      global_insertion_points.reserve(this->inserted_this_step);
+      insertion_points_on_proc.reserve(this->inserted_this_step_this_proc);
+
+      this->assign_insertion_points(global_insertion_points,
+                                    dem_parameters.insertion_info);
+
+      // Distributing the insertion points between the processors
+      if (this_mpi_process != (n_mpi_process - 1))
+        {
+          int begin_element =
+            this_mpi_process * this->inserted_this_step_this_proc;
+          int end_element =
+            (this_mpi_process + 1) * this->inserted_this_step_this_proc - 1;
+          insertion_points_on_proc = std::vector<Point<dim>>(
+            global_insertion_points.cbegin() + begin_element,
+            global_insertion_points.cbegin() + end_element + 1);
+        }
+      else
+        {
+          int begin_element =
+            this->inserted_this_step - this->inserted_this_step_this_proc;
+          insertion_points_on_proc =
+            std::vector<Point<dim>>(global_insertion_points.cbegin() +
+                                      begin_element,
+                                    global_insertion_points.cend());
+        }
```

Comment:
Wait, so all processor calculate the entire std::vector of points, but then only takes a sub-part of it? Why? This seems expensive for no reason. Imagine you have 1M particles, but 1000 core. Each core will create a vector of 1M points just to use 1K points of it? You could calculate only the points necessary by obtaining the i,j,k id from the global particle number using modulo and integer division... It would be much cleaner than to carry the global vector all around.

Reply:
revised